### ISSUE TITLE
Allow instantiation of internal classes via stringified class name 

### ISSUE THREAD

**Issue by WaynePhillipsEA** [WaynePhillipsEA/twinbasic#15]
**Is your feature request related to a problem? Please describe.**
It could be useful to be able to instantiate internal twinBASIC classes via their qualified, stringified names, much like CreateObject(), but without requiring typelibrary registration, and supporting constructor arguments.

**Describe the solution you'd like**
The best solution might be to offer a small API for enumerating internal classes and their (multiple) constructors, and allowing the developer to Invoke() the constructor.  A tiny reflection-like API.

**Additional context**
This would be particularly useful for implementing an IoC container.

**Source**
Rubberduck chat [mansellan]

**Comment by wqweto** [comment]
Best would be to augment `CreateObject` to be able to pass private class ProgID e.g. this
```
Set o = CreateObject("Project1.Class1")
```
. . . should work in Std-EXE projects (all classes being private).

This is an omission in VB6 and so now we have to resort to hacks like this [NameBasedObjectFactory.bas](https://www.vbforums.com/showthread.php?834231-Instantiate-internal-class-object-with-name-in-string&p=5082493&viewfull=1#post5082493) instead of just using a more streamlined `CreateObject` built-in function.

**Comment by bclothier** [comment]
Just to note: I would not want to see any new features that depends on stringification to work around some limitations or such. Stringifying is inherently fragile because renaming can lead to runtime errors. I believe that a language should make it easy to refactor code, and stringification actively hinder this. 

To make it a compile-time error and thus easy to refactor, we can consider implementing a equivalent of `nameof()` function. That does not help in the case where we don't know the types at the compile-time, though.

**Comment by wqweto** [comment]
`CreateObject` on private classes would be helpful when the ProgID comes from an external application on the command-line for instance.

This is just to skip the giant switch one has to implement for instantiating the classes which might even be not possible to write if the .vbp project file is dynamicly generated (like in this [make.bat](https://github.com/wqweto/VbPeg/blob/master/test/Runner/make.bat) file).

**Comment by retailcoder** [comment]
In the realm of relatively simple things that have great impact, some `Reflection.GetMemberNames` function that takes a *type* (or a type *name*?) and returns an array containing the names of each (public) member in the specified type would be amazing all by itself, even if it only ever works with `Enum` types.

**Comment by mansellan** [comment]
A full solution for reflection would be awesome, and open up new fields (dynamic proxies, anyone?), but would require type metadata to be emitted for runtime retrieval (more info than in the typelib, because overloads, generics, and other tB goodness).

@WaynePhillipsEA , is this feasible (even if not immediately)?

**Comment by WaynePhillipsEA** [comment]
Yes, I would think so.  I'm actually touching this area of tB next week because we need to include the extended type data in order to implement the IDispatch class implementation (for late bound support, including runtime overload resolution etc).  So.... I'll get back to you on this, and I'll have this in the back of my mind whilst doing the IDispatch work.

**Comment by wqweto** [comment]
I thought that generally in VB6 and ATL/MFC an IDispatch implementation for a dual interface is forwarding IDispatch methods to system provided helper functions e.g. IDispatch::Invoke to [DispInvoke](https://docs.microsoft.com/en-us/windows/win32/api/oleauto/nf-oleauto-dispinvoke) which is using the already available early-bound interface vtable + dual-interface typeinfo from the typelib to handle all pesky edge-cases, for instance Variants with VT_BYREF wrapping VT_I4 (passed ByRef), DISPATCH_PROPERTYGET | DISPATCH_METHOD combined flags from scripting clients and what not.

Not using OS provided helpers risks leading to bizarre  bugs like in StdPicture::Render method which bombs on VT_BYREF arguments with strange errors in VB6 and also XMLHTTP::Send mthod which has troubles with VT_BYREF too.

**Comment by WaynePhillipsEA** [comment]
@wqweto Usually, yes.  But we can't do that as we need to support tB's overloaded members, and DispGetIDsOfNames/DispInvoke etc cannot handle them as ordinary COM doesn't support them.

Additionally, by implementing a custom version, it will be ready for when we go cross-platform.  The issues you raise are good points and we are aware of them.

**Comment by wqweto** [comment]
I see and this makes sense.

For cross-platform I would personally prefer to re-implement DispInvoke in the compatibility layer on non-Windows platforms but enhanced DispInvoke equivalent implementation with non-COM features is probably a good decision too although way too much effort -- you now have to devise enhanced typeinfo for interface to be stored in enhanced typelibs too and provide conversion from enhanced to standard typelibs for the public interfaces in Ax-DLLs.

**Comment by WaynePhillipsEA** [comment]
@wqweto Don't get dragged down in the detail ;)    This has all been thought through for a very long time, and many parts of the puzzle are already in place.  I can assure you, it won't be a problem. 👍
